### PR TITLE
UCP: UCT/CM based wireup and close protocol draft for review

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
UCT/CM based UCP wireup and close protocol draft for review

@yosefe @alinask

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11492/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 1 of 1 workers (click for details)</summary>

**Note:** the logs will be deleted after 31-Jul-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/b2c24d2b1044f9450217ed562d1534fc/raw/17064fe6c130f13739b238ed323f5bbaeb92978e/_main) | :x: FAILURE

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11502/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 1 of 1 workers (click for details)</summary>

**Note:** the logs will be deleted after 31-Jul-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/331e921e6f855283382f8426fb288d79/raw/27b1ea47a4d4680c29911e7d91ef789735677f92/_main) | :x: FAILURE

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11537/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 1 of 1 workers (click for details)</summary>

**Note:** the logs will be deleted after 02-Aug-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/9273ec462d2ec7da74cfafc263f78697/raw/386c9fa2890dbe976dc9ce50c7cf62e94ce58890/_main) | :x: FAILURE

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11539/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 1 of 1 workers (click for details)</summary>

**Note:** the logs will be deleted after 02-Aug-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/ef9b22485a9ad90fbdc566c546d97a06/raw/fb922a6607c33258596b161aa99aaba059930b19/_main) | :x: FAILURE

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11577/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 1 of 1 workers (click for details)</summary>

**Note:** the logs will be deleted after 05-Aug-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/07723270e64286a68cda60a930ae734b/raw/c5e476e079166e6a619939d7741ec60d0bfe567d/_main) | :x: FAILURE

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -247,7 +247,14 @@ static ucs_config_field_t ucp_config_table[] = {
    "of all entities which connect to each other are the same.",
    ucs_offsetof(ucp_config_t, ctx.unified_mode), UCS_CONFIG_TYPE_BOOL},
 
-  {NULL}
+  {"SOCKADDR_CLOSE_PROTO", "n",
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: should be used by default if available

File: src/ucp/core/ucp_ep.c [hunk 2]
```diff
@@ -219,6 +207,225 @@ ucp_ep_adjust_params(ucp_ep_h ep, const ucp_ep_params_t *params)
...
+    ucs_assert(ucp_ep_get_connected_ep(ucp_ep) == NULL);
+
+    new_key = ucp_ep_config(ucp_ep)->key;
+    if (new_key.num_lanes == UCP_MAX_LANES) {
+        return UCS_ERR_EXCEEDS_LIMIT;
+    }
+
+    new_key.connected_lane                  = new_key.num_lanes++;
+    ucp_ep->uct_eps[new_key.connected_lane] = uct_ep;
+    return ucp_worker_get_ep_config(ucp_ep->worker, &new_key, 0,
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: must be careful changing cfg_index of an ep which can already have operations in progress

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -219,6 +207,225 @@ ucp_ep_adjust_params(ucp_ep_h ep, const ucp_ep_params_t *params)
...
+
+    lane_params.user_data                  = ep;
+    lane_params.conn_request               = params->conn_request->uct_req;
+    lane_params.sockaddr_cb_flags          = UCT_CB_FLAG_ASYNC;
+    lane_params.sockaddr_pack_cb           = ucp_wireup_sockaddr_priv_pack_cb;
+    lane_params.sockaddr_connect_cb.server = ucp_wireup_sockaddr_server_connect_cb;
+    lane_params.disconnect_cb              = ucp_ep_sockaddr_disconnect_cb;
+
+    ucs_assert(worker->num_cms == 1);
+    for (i = 0; i < worker->num_cms; ++i) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: we can support as many CMs as needed, as long as it's 1 :)

File: src/ucp/core/ucp_ep.c [hunk 4]
```diff
@@ -219,6 +207,225 @@ ucp_ep_adjust_params(ucp_ep_h ep, const ucp_ep_params_t *params)
...
+        status = ucp_wireup_ep_connect_to_ep(&wireup_ep->super.super,
+                                             remote_address->address_list[0].dev_addr,
+                                             remote_address->address_list[0].ep_addr);
+    } else {
+        uct_ep_params_t params;
+        params.field_mask = UCT_EP_PARAM_FIELD_IFACE    |
+                            UCT_EP_PARAM_FIELD_DEV_ADDR |
+                            UCT_EP_PARAM_FIELD_IFACE_ADDR;
+        params.iface      = wireup_ep->sockaddr_wiface->iface;
+        params.dev_addr   = remote_address->address_list[0].dev_addr;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: need to think about some best-transport selection logic

File: src/ucp/core/ucp_ep.c [hunk 5]
```diff
@@ -449,13 +673,28 @@ ucs_status_t ucp_ep_create_accept(ucp_worker_h worker,
...
+        ucs_assert(ucp_worker_close_proto(worker));
+        ucs_assert(client_data->addr_mode == UCP_WIREUP_SOCKADDR_CD_LOCAL_ADDR);
+        ucs_assert(remote_address.address_count == 1);
+        remote_address.address_list[0].dev_addr = conn_request->remote_dev_addr;
+        params.field_mask  |= UCP_EP_PARAM_FIELD_CONN_REQUEST;
+        params.conn_request = conn_request;
+        status = ucp_ep_create_sockaddr_connected(worker, &params,
+                                                  &remote_address, ep_p);
+        (*ep_p)->flags |= UCP_EP_FLAG_LISTENER;
+        ucp_ep_ext_gen(*ep_p)->listener = conn_request->ucp_listener;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: why do we need to save the listener?

File: src/ucp/core/ucp_ep.inl [hunk 6]
```diff
@@ -193,15 +204,16 @@ static inline void ucp_ep_flush_state_reset(ucp_ep_h ep)
...
-    ucs_assert(!(ep->flags & UCP_EP_FLAG_FLUSH_STATE_VALID) ||
-               ((flush_state->send_sn == 0) &&
-                (flush_state->cmpl_sn == 0) &&
-                ucs_queue_is_empty(&flush_state->reqs)));
-
-    flush_state->send_sn = 0;
-    flush_state->cmpl_sn = 0;
-    ucs_queue_head_init(&flush_state->reqs);
-    ep->flags |= UCP_EP_FLAG_FLUSH_STATE_VALID;
+    if (!(ep->flags & UCP_EP_FLAG_FLUSH_STATE_VALID)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.inl]: looks like rebase issue

File: src/ucp/wireup/wireup_ep.c [hunk 7]
```diff
@@ -573,6 +573,227 @@ ssize_t ucp_wireup_ep_sockaddr_fill_private_data(void *arg, const char *dev_name
...
+{
+    const size_t    dev_name_length = strlen(dev_name);
+    ucp_rsc_index_t num_tls;
+    ucp_rsc_index_t iface_idx;
+    ucp_rsc_index_t tl_rsc_idx;
+
+    *wiface_p = NULL;
+    num_tls = worker->context->num_tls;
+    for (tl_rsc_idx = 0; tl_rsc_idx < num_tls; ++tl_rsc_idx) {
+        if (!strncmp(worker->context->tl_rscs[tl_rsc_idx].tl_rsc.dev_name,
```
Comment (yosefe) [inline src/ucp/wireup/wireup_ep.c]: why not strcmp?

File: src/ucp/wireup/wireup_ep.c [hunk 8]
```diff
@@ -573,6 +573,227 @@ ssize_t ucp_wireup_ep_sockaddr_fill_private_data(void *arg, const char *dev_name
...
+    ucp_worker_iface_t *wiface;
+    uct_cm_attr_t cm_attr;
+    uct_ep_params_t tl_ep_params;
+    ucp_rsc_index_t cm_idx;
+    ucs_status_t status;
+
+    UCS_ASYNC_BLOCK(&ep->worker->async);
+
+    for (cm_idx = 0; cm_idx < ep->worker->num_cms; ++cm_idx) {
+        status = uct_cm_query(ep->worker->cms[cm_idx], &cm_attr);
```
Comment (yosefe) [inline src/ucp/wireup/wireup_ep.c]: cm attr should be cached on context

**Comment by evgeny-leksikov** [comment]
outdated and broken up to smaller PRs

**Pull request closed without merging**